File: QuickSort/Python/QuickSort_Correctness.py
# QuickSort_Correctness.py
"""
Correctness test harness for Python sorting algorithms.

Assumes each algorithm is a function:
    def quicksort(a: list[int]) -> None
"""

from __future__ import annotations
from typing import Callable, Dict, List, Tuple
import random
import traceback

# Import Algorithms

from QuickSort_ChatGPT import quicksort as quicksort_chatgpt
from QuickSort_Claude import quicksort as quicksort_claude
from QuickSort_DeepSeek import quicksort as quicksort_deepseek
from QuickSort_Gemini import quicksort as quicksort_gemini

SortFunc = Callable[[List[int]], None]

ALGORITHMS: Dict[str, SortFunc] = {
    "Chatgpt" : quicksort_chatgpt,
    "Claude" : quicksort_claude,
    "Deepseek" : quicksort_deepseek,
    "Gemini" : quicksort_gemini,
}

# Test Case Generators

def gen_random(n: int) -> List[int]:
    return [random.randint(-10**6, 10**6) for _ in range(n)]

def gen_sorted(n: int) -> List[int]:
    arr = gen_random(n)
    arr.sort()
    return arr

def gen_reversed(n: int) -> List[int]:
    arr = gen_random(n)
    arr.sort(reverse=True)
    return arr

def gen_nearly_sorted(n: int) -> List[int]:
    """Start sorted, then do a few random swaps."""
    arr = gen_sorted(n)
    if n <= 1:
        return arr
    num_swaps = max(1, n // 100)  
    for _ in range(num_swaps):
        i = random.randrange(n)
        j = random.randrange(n)
        arr[i], arr[j] = arr[j], arr[i]
    return arr

def gen_few_values(n: int) -> List[int]:
    """Many duplicates: values from a small range."""
    return [random.randint(0, 9) for _ in range(n)]

EDGE_CASES: List[List[int]] = [
    [],
    [1],
    [1, 1, 1],
    [0, -1, 5, -1],
    [2, 1],
    [2, 2, 1, 1],
    [5, 4, 3, 2, 1],
]

DISTRIBUTIONS: Dict[str, Callable[[int], List[int]]] = {
    "random": gen_random,
    "sorted": gen_sorted,
    "reversed": gen_reversed,
    "nearly_sorted": gen_nearly_sorted,
    "few_values": gen_few_values,
}

SIZES = [0, 1, 2, 5, 10, 100, 1000, 5000]
CASES_PER_COMBO = 20  # number of random arrays per (distribution, size)

# Run Tests

def run_correctness_tests() -> None:

    random.seed(1)  # for reproducibility

    for algo_name, sort_func in ALGORITHMS.items():
        print("=" * 70)
        print(f"Testing algorithm: {algo_name}")
        total_tests = 0
        total_failures = 0
        total_exceptions = 0

        # Edge cases first
        print("  Edge cases:")
        for idx, case in enumerate(EDGE_CASES):
            total_tests += 1
            arr = list(case)
            expected = sorted(case)
            try:
                sort_func(arr)
                if arr != expected:
                    total_failures += 1
                    print(f"    FAIL edge_case[{idx}]: input={case}, got={arr}, expected={expected}")
            except Exception as e:
                total_exceptions += 1
                #print(f"    EXCEPTION edge_case[{idx}]: {e}")
                #traceback.print_exc(limit=1)

        # Randomized distributions
        print("  Randomized distributions:")
        for dist_name, generator in DISTRIBUTIONS.items():
            dist_tests = 0
            dist_failures = 0
            dist_exceptions = 0

            for n in SIZES:
                for _ in range(CASES_PER_COMBO):
                    arr = generator(n)
                    expected = sorted(arr)
                    test_input = list(arr)  # keep original for debugging
                    total_tests += 1
                    dist_tests += 1
                    try:
                        sort_func(arr)
                        if arr != expected:
                            total_failures += 1
                            dist_failures += 1
                            print(
                                f"    FAIL {dist_name}, n={n}: "
                                f"input(sample)={test_input[:10]}, "
                                f"got(sample)={arr[:10]}, expected(sample)={expected[:10]}"
                            )
                    except Exception as e:
                        total_exceptions += 1
                        dist_exceptions += 1
                        #print(f"    EXCEPTION {dist_name}, n={n}: {e}")
                        #traceback.print_exc(limit=1)

            print(
                f"    {dist_name}: tests={dist_tests}, "
                f"failures={dist_failures}, exceptions={dist_exceptions}"
            )

        print("-" * 70)
        print(
            f"Summary for {algo_name}: "
            f"tests={total_tests}, failures={total_failures}, exceptions={total_exceptions}"
        )
        print()

    print("All correctness tests completed.")


if __name__ == "__main__":
    run_correctness_tests()

File: QuickSort/Rust/src/bin/quicksort_correctness.rs
// src/bin/quicksort_correctness.rs
//
// Correctness test harness for Rust sorting algorithms.
//
// Run with:
//    cargo run --bin quicksort_correctness
//
// This assumes each assistant implementation is in the format
//    pub fn quicksort(a: &mut [i32]) { ... }

use std::collections::HashMap;
use std::panic::{catch_unwind, AssertUnwindSafe};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

type SortFn = fn(&mut [i32]);

// Import Algorithms

mod quicksort_chatgpt;
mod quicksort_claude;
mod quicksort_deepseek;
mod quicksort_gemini;


use quicksort_chatgpt::quicksort as quicksort_chatgpt_fn;
use quicksort_claude::quicksort as quicksort_claude_fn;
use quicksort_deepseek::quicksort as quicksort_deepseek_fn;
use quicksort_gemini::quicksort as quicksort_gemini_fn;

fn algorithms() -> HashMap<&'static str, SortFn> {
    HashMap::from([
        ("Chatgpt", quicksort_chatgpt_fn as SortFn),
        ("Claude", quicksort_claude_fn as SortFn),
        ("Deepseek", quicksort_deepseek_fn as SortFn),
        ("Gemini", quicksort_gemini_fn as SortFn),
    ])
}

// Test Case Generators

fn gen_random(n: usize, rng: &mut StdRng) -> Vec<i32> {
    (0..n)
        .map(|_| rng.gen_range(-1_000_000..=1_000_000))
        .collect()
}

fn gen_sorted(n: usize, rng: &mut StdRng) -> Vec<i32> {
    let mut v = gen_random(n, rng);
    v.sort();
    v
}

fn gen_reversed(n: usize, rng: &mut StdRng) -> Vec<i32> {
    let mut v = gen_random(n, rng);
    v.sort_by(|a, b| b.cmp(a));
    v
}

fn gen_nearly_sorted(n: usize, rng: &mut StdRng) -> Vec<i32> {
    let mut v = gen_sorted(n, rng);
    if n <= 1 {
        return v;
    }
    let num_swaps = (n / 100).max(1); // ~1%
    for _ in 0..num_swaps {
        let i = rng.gen_range(0..n);
        let j = rng.gen_range(0..n);
        v.swap(i, j);
    }
    v
}

fn gen_few_values(n: usize, rng: &mut StdRng) -> Vec<i32> {
    (0..n).map(|_| rng.gen_range(0..=9)).collect()
}

const DIST_NAMES: &[&str] = &["random", "sorted", "reversed", "nearly_sorted", "few_values"];

fn make_array(dist: &str, n: usize, rng: &mut StdRng) -> Vec<i32> {
    match dist {
        "random" => gen_random(n, rng),
        "sorted" => gen_sorted(n, rng),
        "reversed" => gen_reversed(n, rng),
        "nearly_sorted" => gen_nearly_sorted(n, rng),
        "few_values" => gen_few_values(n, rng),
        _ => panic!("unknown distribution {dist}"),
    }
}

fn edge_cases() -> Vec<Vec<i32>> {
    vec![
        vec![],
        vec![1],
        vec![1, 1, 1],
        vec![0, -1, 5, -1],
        vec![2, 1],
        vec![2, 2, 1, 1],
        vec![5, 4, 3, 2, 1],
    ]
}

const SIZES: &[usize] = &[0, 1, 2, 5, 10, 100, 1000, 5000];
const CASES_PER_COMBO: usize = 20;

// Run Tests

fn run_correctness_tests() {
    let algos = algorithms();
    if algos.is_empty() {
        eprintln!("No algorithms registered.");
        return;
    }

    let mut rng = StdRng::seed_from_u64(1);

    for (name, sort_fn) in algos {
        println!("{}", "=".repeat(70));
        println!("Testing algorithm: {name}");
        let mut total_tests = 0usize;
        let mut total_failures = 0usize;
        let mut total_exceptions = 0usize;

        println!("  Edge cases:");
        for (idx, case) in edge_cases().into_iter().enumerate() {
            total_tests += 1;
            let mut arr = case.clone();
            let mut expected = case.clone();
            expected.sort();

            let res = catch_unwind(AssertUnwindSafe(|| {
                sort_fn(&mut arr);
            }));

            match res {
                Ok(_) => {
                    if arr != expected {
                        total_failures += 1;
                        println!(
                            "    FAIL edge_case[{idx}]: input={:?}, got={:?}, expected={:?}",
                            case, arr, expected
                        );
                    }
                }
                Err(_) => {
                    total_exceptions += 1;
                    println!("    PANIC edge_case[{idx}]");
                }
            }
        }

        println!("  Randomized distributions:");
        for &dist in DIST_NAMES {
            let mut dist_tests = 0usize;
            let mut dist_failures = 0usize;
            let mut dist_exceptions = 0usize;

            for &n in SIZES {
                for _ in 0..CASES_PER_COMBO {
                    total_tests += 1;
                    dist_tests += 1;

                    let input = make_array(dist, n, &mut rng);
                    let mut arr = input.clone();
                    let mut expected = input.clone();
                    expected.sort();

                    let res = catch_unwind(AssertUnwindSafe(|| {
                        sort_fn(&mut arr);
                    }));

                    match res {
                        Ok(_) => {
                            if arr != expected {
                                total_failures += 1;
                                dist_failures += 1;
                                println!(
                                    "    FAIL {dist}, n={n}: input(sample)={:?}, got(sample)={:?}, expected(sample)={:?}",
                                    &input.get(0..10).unwrap_or(&input[..]),
                                    &arr.get(0..10).unwrap_or(&arr[..]),
                                    &expected.get(0..10).unwrap_or(&expected[..]),
                                );
                            }
                        }
                        Err(_) => {
                            total_exceptions += 1;
                            dist_exceptions += 1;
                            println!("    PANIC {dist}, n={n}");
                        }
                    }
                }
            }

            println!(
                "    {dist}: tests={dist_tests}, failures={dist_failures}, panics={dist_exceptions}"
            );
        }

        println!("{}", "-".repeat(70));
        println!(
            "Summary for {name}: tests={total_tests}, failures={total_failures}, panics={total_exceptions}\n"
        );
    }

    println!("All correctness tests completed.");
}

fn main() {
    run_correctness_tests();
}
